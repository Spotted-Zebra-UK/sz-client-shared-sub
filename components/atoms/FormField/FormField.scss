@import '../../../variables';

.FormField {
  &__Label__Text {
    font-size: $p2-font-size;
    line-height: $small-line-height;
    text-align: left;
  }

  &__Label__Text--HasError {
    color: $purple;
  }

  &__Error {
    color: $purple;
    font-size: $p2-font-size;
    min-height: 16px;
    line-height: $small-line-height;
    margin-top: 8px;
    text-align: left;
  }
}
