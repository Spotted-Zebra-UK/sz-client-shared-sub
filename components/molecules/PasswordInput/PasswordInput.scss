@import '../../../variables';

.PasswordInput {
  border-bottom: 2px solid $black;
  display: flex;

  input {
    border: 0;
    font-size: $p1-font-size;
    height: 36px;
    line-height: $medium-line-height;
    width: calc(100% - 22px);
    font-size: 14px;
  }

  &__Button {
    border: 0;
    background-color: transparent;
    cursor: pointer;

    svg {
      width: 22px;

      path {
        fill: rgba(0, 0, 0, 0.16);
      }
    }
  }
}
