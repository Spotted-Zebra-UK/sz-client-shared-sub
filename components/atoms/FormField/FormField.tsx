import './FormField.scss';
import React, { FunctionComponent } from 'react';

interface IFormField {
  error?: string;
  fieldName: string;
  label?: string;
}

const FormField: FunctionComponent<IFormField> = ({
  children,
  error,
  label,
  fieldName,
}) => {
  return (
    <div className={`FormField FormField--${fieldName}`}>
      <label className="FormField__Label" htmlFor={fieldName}>
        {label ? (
          <div
            className={`FormField__Label__Text${
              error ? ' FormField__Label__Text--HasError' : ''
            }`}
          >
            {label}
          </div>
        ) : null}
        {children}
        <div className="FormField__Error">{error}</div>
      </label>
    </div>
  );
};

export default FormField;
