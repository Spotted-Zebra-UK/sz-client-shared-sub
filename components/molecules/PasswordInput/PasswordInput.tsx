import './PasswordInput.scss';
import React, { FC, useState } from 'react';
import { ReactComponent as EyeIcon } from '../../../icons/Eye.svg';
import { ReactComponent as EyeSolidIcon } from '../../../icons/EyeSolid.svg';
import Button from '../../atoms/Button/Button';

interface IPasswordInput {
  name: string;
  onChange: (value: string, name: string) => void;
  placeholder?: string;
  value: string;
}

const PasswordInput: FC<IPasswordInput> = ({
  value,
  onChange,
  placeholder,
  name,
}) => {
  const [isVisible, setIsVisible] = useState(false);

  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    onChange(event.target.value, name);
  };

  const handleVisibilityChange = () => {
    setIsVisible(prevIsVisible => !prevIsVisible);
  };

  return (
    <div className="PasswordInput">
      <input
        onChange={handleChange}
        name={name}
        placeholder={placeholder}
        value={value}
        type={isVisible ? 'text' : 'password'}
      />
      <Button
        onClick={handleVisibilityChange}
        className="PasswordInput__Button"
      >
        {isVisible ? <EyeSolidIcon /> : <EyeIcon />}
      </Button>
    </div>
  );
};

export default PasswordInput;
