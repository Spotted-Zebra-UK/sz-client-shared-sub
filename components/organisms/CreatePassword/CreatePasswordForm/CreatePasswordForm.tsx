/* eslint-disable @typescript-eslint/indent */
import './CreatePasswordForm.scss';
import React, { FC, useState } from 'react';
import validate from '../../../../helpers/validate';
import { TFormFieldValue } from '../../../../interfaces/form';
import FormField from '../../../atoms/FormField/FormField';
import PasswordInput from '../../../molecules/PasswordInput/PasswordInput';
import SquareButton from '../../../molecules/SquareButton/SquareButton';

interface ICreatePasswordFormValues {
  password: string;
}

interface ICreatePasswordFormErrors {
  password: string[];
}

interface ICreatePasswordFormForm {
  onSubmit: (password: string) => void;
}

const CreatePasswordForm: FC<ICreatePasswordFormForm> = props => {
  const [values, setValues] = useState<ICreatePasswordFormValues>({
    password: '',
  });
  const [errors, setErrors] = useState<ICreatePasswordFormErrors | undefined>();

  const CreatePasswordFormValidate = () => {
    return validate(values, {
      password: {
        length: { minimum: 5 },
      },
    });
  };

  const handleChange = (value: TFormFieldValue, name: string) => {
    setValues(prevValues => ({ ...prevValues, [name]: value }));
    setErrors(prevErrors =>
      prevErrors ? { ...prevErrors, [name]: [] } : prevErrors
    );
  };

  const handleSubmit = (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    const newErrors = CreatePasswordFormValidate();
    if (newErrors) {
      setErrors(newErrors);
    } else {
      props.onSubmit(values.password);
    }
  };

  return (
    <form className="CreatePasswordForm" onSubmit={handleSubmit}>
      <div className="CreatePasswordForm__Fields">
        <FormField
          error={errors && errors.password && errors.password.join(' ')}
          fieldName="password"
          label="Password"
        >
          <PasswordInput
            name="password"
            onChange={handleChange}
            placeholder="Password"
            value={values.password}
          />
        </FormField>
      </div>
      <SquareButton type="submit">Confirm</SquareButton>
    </form>
  );
};

export default CreatePasswordForm;
