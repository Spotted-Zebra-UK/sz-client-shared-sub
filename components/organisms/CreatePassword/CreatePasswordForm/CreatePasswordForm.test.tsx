import { shallow } from 'enzyme';
import FormField from '../../../atoms/FormField/FormField';
import PasswordInput from '../../../molecules/PasswordInput/PasswordInput';
import CreatePasswordForm from './CreatePasswordForm';

describe('CreatePasswordForm', () => {
  let onSubmit: (password: string) => void;

  beforeEach(() => {
    onSubmit = jest.fn();
  });

  // it('should render correctly by default', () => {
  //   const wrapper = shallow(<CreatePasswordForm onSubmit={onSubmit} />);

  //   expect(toJson(wrapper)).toMatchSnapshot();
  //   expect(wrapper.find(FormField).props().error).toBeUndefined();
  //   expect(wrapper.find(PasswordInput).props().value).toBe('');
  // });

  it('should change input value on input change', () => {
    const passwordValue = 'test-passwordValue';
    const wrapper = shallow(<CreatePasswordForm onSubmit={onSubmit} />);

    wrapper.find(PasswordInput).props().onChange(passwordValue, 'password');
    expect(wrapper.find(PasswordInput).props().value).toBe(passwordValue);
  });

  it('should call provided onSubmit', () => {
    const passwordValue = 'test-passwordValue';
    const wrapper = shallow(<CreatePasswordForm onSubmit={onSubmit} />);

    wrapper.find(PasswordInput).props().onChange(passwordValue, 'password');
    expect(wrapper.find(PasswordInput).props().value).toBe(passwordValue);

    const eventMock = {
      preventDefault: jest.fn(),
    };

    wrapper.find('form').simulate('submit', eventMock);

    expect(eventMock.preventDefault).toHaveBeenCalled();
    expect(onSubmit).toHaveBeenCalledWith(passwordValue);
  });

  it('should show error and not call onSubmit if submit too short password', () => {
    const wrapper = shallow(<CreatePasswordForm onSubmit={onSubmit} />);

    const eventMock = {
      preventDefault: jest.fn(),
    };

    wrapper.find('form').simulate('submit', eventMock);

    expect(eventMock.preventDefault).toHaveBeenCalled();
    expect(wrapper.find(FormField).props().error).toBe(
      'Password is too short (minimum is 5 characters)'
    );
    expect(onSubmit).not.toHaveBeenCalled();
  });

  it('should hide error if error is visible and input changed', () => {
    const passwordValue = 'test-passwordValue';
    const wrapper = shallow(<CreatePasswordForm onSubmit={onSubmit} />);

    const eventMock = {
      preventDefault: jest.fn(),
    };

    wrapper.find('form').simulate('submit', eventMock);

    expect(wrapper.find(FormField).props().error).toBe(
      'Password is too short (minimum is 5 characters)'
    );

    wrapper.find(PasswordInput).props().onChange(passwordValue, 'password');
    expect(wrapper.find(FormField).props().error).toBe('');
  });
});
