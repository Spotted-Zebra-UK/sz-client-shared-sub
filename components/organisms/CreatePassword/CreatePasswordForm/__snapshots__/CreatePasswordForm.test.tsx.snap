// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreatePasswordForm should render correctly by default 1`] = `
<form
  className="CreatePasswordForm"
  onSubmit={[Function]}
>
  <div
    className="CreatePasswordForm__Fields"
  >
    <FormField
      fieldName="password"
      label="Password"
    >
      <PasswordInput
        name="password"
        onChange={[Function]}
        placeholder="Password"
        value=""
      />
    </FormField>
  </div>
  <SquareButton
    type="submit"
  >
    Confirm
  </SquareButton>
</form>
`;
